### DISCUSSION TITLE
What is the need to add the "fn" attachment?

### DISCUSSION THREAD

**Discussion by boludoz** [carbon-language/carbon-lang#3289] [Getting started]
What is the need to add the "fn" add-on and look like Rust instead of C? Why complicate things?

**Comment by boludoz** [comment]
instead:
```
fn Main() -> i32 {
  // A dynamically sized array, like `std::vector`.
  var circles: Array(Circle) = ({.r = 1.0}, {.r = 2.0});
  // Implicitly constructs `Slice` from `Array`.
  PrintTotalArea(circles);
  return 0;
}

```
Could be:
```
i32 Main() {
  // A dynamically sized array, like `std::vector`.
  var circles: Array(Circle) = ({.r = 1.0}, {.r = 2.0});
  // Implicitly constructs `Slice` from `Array`.
  PrintTotalArea(circles);
  return 0;
}

```

**Comment by boludoz** [comment]
Adding "->" and "fn" in real life is a waste of time and makes life difficult for people who are just starting out.

The objective of a language is that it should be as versatile and direct as possible in every sense.

**Comment by boludoz** [comment]
I also propose that the "auto" type be added. With automatic type conversion in a standardized way, which allows recycling of functions to obtain more compact codes.

**Comment by KateGregory** [comment]
> What is the need to add the "fn" add-on and look like Rust instead of C? Why complicate things?

This is discussed in https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0438.md#function-keyword . Just using the return type to say "here comes a function declaration" may feel less complicated to you because you're familiar with it, but that doesn't mean it's simpler for everyone. It makes the parsing (for the compiler) harder. It may make reading the code harder for people who are not coming from a C or C++ background and expect an introducer like fn to explicitly mark the start of a function.

**Reply by boludoz** [reply]
Is it necessary to declare the type to be returned? Because if it were not necessary that would be good.

**Reply by geoffromer** [reply]
You can omit the return type if it's `()`, or use `auto` to have the return type be deduced. See https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/functions.md
